darwin.Libsystem: export i386 symbols

veprbl opened the pull request:
Current darwin.Libsystem loses i386 symbols when doing reexporting of
symbols from some of the libraries. This effectively prevents users from
compiling with -m32. This commit exports i386 symbols from
libsystem_{c,kernel}.dylib into libSystem.B.dylib.  This does not
address a single symbol from libSystem_internal.dylib.

The $out/lib/system/libsystem_{c,kernel}.dylib will still come as x86_64
only, but that doesn't seem to be important.

###### Motivation for this change

* Advance in fixing wine for Darwin (#8849)
* Might fix SML/NJ on Darwin (#16018)

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

veprbl committed:
darwin.Libsystem: export i386 symbols

Current darwin.Libsystem loses i386 symbols when doing reexporting of
symbols from some of the libraries. This effectively prevents users from
compiling with -m32. This commit exports i386 symbols from
libsystem_{c,kernel}.dylib into libSystem.B.dylib.  This does not
address a single symbol in libSystem_internal.dylib.

The $out/lib/system/libsystem_{c,kernel}.dylib will still come as x86_64
only, but that doesn't seem to be important.

veprbl commented:
cc @copumpkin @LnL7

veprbl commented:
Fixing wine would also require libresolv being built as universal binary, which can't be done unless bootstrap tools are updated and then pkgs/stdenv/darwin/unpack-bootstrap-tools.sh will also need an update.

copumpkin commented:
I see nothing wrong with this, but I'm also surprised that anyone cares about 32-bit anything on macOS 😄 the only Intel Macs that don't support 64-bit code existed for a year or two back in 2006 or something.

LnL7 commented:
Isn't it possible to disable the universal parts of the build? I'm building the changes to make sure nothing obvious breaks.
How did you generate the i686 symbols?

veprbl commented:
@copumpkin Oh, I'm with you on this. But wine seems to need that (Windows had 32-bit for couple of decades) and some legacy software may require that too. This probably shouldn't bloat things too much (extra bloat might happen in clang as it seems to cleverly detect universal libSystem and builds an i386 runtime as well).

@LnL7 Sorry, don't understand which build you refer to. win32 seems to need 32 bit binaries to be natively be called from 32 bit windows software, I don't know many details, though. About your second point: I had exactly the same question about x86_64 symbols that were already there ;) In my commit I add gen_symb_lists.sh script, which I used to generate the i386 symbols. But it also generates x86_64 symbols and when I run it on my 10.11.6 system it suggests that a few more can be added there.

aflatter commented:
Another motivation: The scrypt ruby gem made me run into the same issue.

copumpkin commented:
I can't really test this right now, but have no fundamental objections if it doesn't break 64-bit stuff.

Main question: can the 32-bit symbols be expressed as a diff from the 64-bit ones, so it's clearer how they differ? Or will the diff be the same size as the full list?

veprbl commented:
@aflatter
I think, it would be good to provide some specific instructions how to reproduce the issue you had. Then we could test it if this lands.

@copumpkin 
Regarding your main question. Initially I was thinking about doing something like that because many entries were quite regular. However there will be exceptions to patterns. For example
```
# ( diff -au system_kernel_symbols_x86_64 system_kernel_symbols_i386;
    diff -au system_c_symbols_x86_64 system_c_symbols_i386 ) \
    | egrep \\\$INODE64\$ -A 2
```
```
 _fstatfs$INODE64 <--------- exception
 _fstatfs64
 _fsync
--
-___opendir2$INODE64
+___opendir2$INODE64$UNIX2003
+___opendir2$UNIX2003
--
 __readdir_unlocked$INODE64 <--------- exception
 __reclaim_telldir
 __seekdir
-__seekdir$INODE64
+__seekdir$INODE64$UNIX2003
+__seekdir$UNIX2003
--
-_fdopendir$INODE64
+_fdopendir$INODE64$UNIX2003
+_fdopendir$UNIX2003
--
 _fts_set$INODE64 <--------- exception
 _ftw
-_ftw$INODE64
+_ftw$INODE64$UNIX2003
+_ftw$UNIX2003
--
-_nftw$INODE64
+_nftw$INODE64$UNIX2003
+_nftw$UNIX2003
--
-_opendir$INODE64
+_opendir$INODE64$UNIX2003
+_opendir$UNIX2003
--
-_rewinddir$INODE64
+_rewinddir$INODE64$UNIX2003
+_rewinddir$UNIX2003
--
-_seekdir$INODE64
+_seekdir$INODE64$UNIX2003
+_seekdir$UNIX2003
--
-_telldir$INODE64
+_telldir$INODE64$UNIX2003
+_telldir$UNIX2003
```
```
( diff -au system_kernel_symbols_x86_64 system_kernel_symbols_i386;
    diff -au system_c_symbols_x86_64 system_c_symbols_i386 ) \
    | egrep \\\$NOCANCEL\$ -A 2
```
```
-_accept$NOCANCEL
+_accept$NOCANCEL$UNIX2003
+_accept$UNIX2003
--
-_aio_suspend$NOCANCEL
+_aio_suspend$NOCANCEL$UNIX2003
+_aio_suspend$UNIX2003
--
-_close$NOCANCEL
+_close$NOCANCEL$UNIX2003
+_close$UNIX2003
--
-_connect$NOCANCEL
+_connect$NOCANCEL$UNIX2003
+_connect$UNIX2003
--
-_fcntl$NOCANCEL
+_fcntl$NOCANCEL$UNIX2003
+_fcntl$UNIX2003
--
-_fsync$NOCANCEL
+_fsync$NOCANCEL$UNIX2003
+_fsync$UNIX2003
--
-_msgrcv$NOCANCEL
+_msgrcv$NOCANCEL$UNIX2003
+_msgrcv$UNIX2003
--
-_msgsnd$NOCANCEL
+_msgsnd$NOCANCEL$UNIX2003
+_msgsnd$UNIX2003
--
-_msync$NOCANCEL
+_msync$NOCANCEL$UNIX2003
+_msync$UNIX2003
--
-_open$NOCANCEL
+_open$NOCANCEL$UNIX2003
+_open$UNIX2003
--
 _openat$NOCANCEL  <--------- exception
@@ -779,7 +821,8 @@
 _pid_suspend
--
-_poll$NOCANCEL
+_poll$NOCANCEL$UNIX2003
+_poll$UNIX2003
--
-_pread$NOCANCEL
+_pread$NOCANCEL$UNIX2003
+_pread$UNIX2003
--
 _pselect$DARWIN_EXTSN$NOCANCEL <--------- exception
-_pselect$NOCANCEL
+_pselect$NOCANCEL$UNIX2003
+_pselect$UNIX2003
--
-_pwrite$NOCANCEL
+_pwrite$NOCANCEL$UNIX2003
+_pwrite$UNIX2003
--
-_read$NOCANCEL
+_read$NOCANCEL$UNIX2003
+_read$UNIX2003
--
-_readv$NOCANCEL
+_readv$NOCANCEL$UNIX2003
+_readv$UNIX2003
--
-_recvfrom$NOCANCEL
+_recvfrom$NOCANCEL$UNIX2003
+_recvfrom$UNIX2003
--
-_recvmsg$NOCANCEL
+_recvmsg$NOCANCEL$UNIX2003
+_recvmsg$UNIX2003
--
 _select$DARWIN_EXTSN$NOCANCEL <--------- exception
-_select$NOCANCEL
+_select$NOCANCEL$UNIX2003
+_select$UNIX2003
--
-_sem_wait$NOCANCEL
+_sem_wait$NOCANCEL$UNIX2003
+_sem_wait$UNIX2003
--
-_sendmsg$NOCANCEL
+_sendmsg$NOCANCEL$UNIX2003
+_sendmsg$UNIX2003
--
-_sendto$NOCANCEL
+_sendto$NOCANCEL$UNIX2003
+_sendto$UNIX2003
--
-_sigsuspend$NOCANCEL
+_sigsuspend$NOCANCEL$UNIX2003
+_sigsuspend$UNIX2003
--
-_waitid$NOCANCEL
+_waitid$NOCANCEL$UNIX2003
+_waitid$UNIX2003
--
-_write$NOCANCEL
+_write$NOCANCEL$UNIX2003
+_write$UNIX2003
--
-_writev$NOCANCEL
+_writev$NOCANCEL$UNIX2003
+_writev$UNIX2003
--
-_creat$NOCANCEL
+_creat$NOCANCEL$UNIX2003
+_creat$UNIX2003
--
-_lockf$NOCANCEL
+_lockf$NOCANCEL$UNIX2003
+_lockf$UNIX2003
--
-_nanosleep$NOCANCEL
+_nanosleep$NOCANCEL$UNIX2003
+_nanosleep$UNIX2003
--
-_pause$NOCANCEL
+_pause$NOCANCEL$UNIX2003
+_pause$UNIX2003
--
-_recv$NOCANCEL
+_recv$NOCANCEL$UNIX2003
+_recv$UNIX2003
--
-_send$NOCANCEL
+_send$NOCANCEL$UNIX2003
+_send$UNIX2003
--
-_sigpause$NOCANCEL
+_sigpause$NOCANCEL$UNIX2003
+_sigpause$UNIX2003
--
-_sleep$NOCANCEL
+_sleep$NOCANCEL$UNIX2003
+_sleep$UNIX2003
--
-_system$NOCANCEL
+_system$NOCANCEL$UNIX2003
+_system$UNIX2003
--
-_tcdrain$NOCANCEL
+_tcdrain$NOCANCEL$UNIX2003
+_tcdrain$UNIX2003
--
-_usleep$NOCANCEL
+_usleep$NOCANCEL$UNIX2003
+_usleep$UNIX2003
--
-_wait$NOCANCEL
+_wait$NOCANCEL$UNIX2003
+_wait$UNIX2003
--
-_waitpid$NOCANCEL
+_waitpid$NOCANCEL$UNIX2003
+_waitpid$UNIX2003
```
You can see quite a few exceptions. And there are some additional differences that don't fall into the regular patterns.

Storing a raw diff would be quite big and probably slightly unpleasant to maintain. Instead of working with diffs I'd personally prefer the usual ```git add -p```, ```git reset -p```. If we consider the "recent" activity related to the ```Libsystem```:

* d8a0307a5d36551c641578a3bf1a7d29103df511 oct10
* 6152c1ea7f9f5beb084386db65805636173528e6 oct13
* 0419452113ebb135907257bb063cb690a4de0b52 jul11
* 99ed13680903033068fd02f49f7c7ff95df7a5e3 dec4

we can tell that the maintenance was mostly about keeping the symbol lists to be a common subset of lists generated for all supported macOS versions. The ```gen_symb_lists.sh``` introduced in this PR combined with commands above along with some documentation should be enough to perform the task.

veprbl committed:
Libsystem: implement 6152c1ea for i386

git show 6152c1ea7f9f5beb084386db65805636173528e6 \
 | sed 's/system_\(c\|kernel\)_symbols/system_\1_symbols_i386/g' \
 | patch -p1 -F3

Original commit message:
Re-enable support for OS X 10.10 for darwin.

veprbl committed:
Libsystem: implement 04194521 for i386

Only changes to system_kernel_symbols need to be applied.

Original commit title:
Fix Darwin stdenv to work on 10.13

veprbl committed:
Libsystem: implement 99ed13680 for i386

git show 99ed13680903033068fd02f49f7c7ff95df7a5e3 \
  | sed 's/system_\(c\|kernel\)_symbols/system_\1_symbols_i386/g' \
  | patch -p1 -F3

Original commit message:
darwin.Libsystem: fix build on 10.13.2

veprbl committed:
Libsystem: remove i386 symbols not present on macOS 10.12.6

This was obtained by running gen_symb_lists.sh and only committing negative parts of the diff. No changes were needed for x86_64 because they were mostly included in d8a0307a5 ('[darwin.stdenv] Fix to work on Sierra').

veprbl committed:
Libsystem: move symbol lists to symbols/

veprbl committed:
Libsystem: generate symbol lists as intersection of the lists for the supported macOS versions

veprbl committed:
Libsystem: drop 10.10 symbols

Returns _pselect$1050 to system_kernel_symbols which is needed by
libresolv. 10.10 has _pselect$1050, but in system_c_symbols.

aflatter commented:
@veprbl Right, so I came up with a minimal repro of my issue:

```shell
tracer:tmp alex$ cat default.nix
let
  pkgs = import <nixpkgs> {};
  ruby = pkgs.ruby_2_3;
  stdenv = pkgs.stdenv;
in
  stdenv.mkDerivation rec {
    name = "repro-scrypt-libsystem";

    buildInputs = [ ruby ];

    shellHook = ''
      export PROJECT_ROOT=$(pwd)
      export GEM_HOME="$PROJECT_ROOT/gems/${ruby.version}"
      export PATH="$GEM_HOME/bin:$PATH"
    '';
  }
tracer:tmp alex$ nix-shell --run 'gem install scrypt'
Building native extensions.  This could take a while...
ERROR:  Error installing scrypt:
	ERROR: Failed to build gem native extension.

    current directory: /Users/alex/tmp/gems/2.3.5/gems/scrypt-3.0.5/ext/scrypt
/nix/store/hm9pigz389gfvr07bid05axzvj93pz1m-ruby-2.3.5/bin/ruby -rubygems /Users/alex/tmp/gems/2.3.5/gems/rake-12.3.0/exe/rake RUBYARCHDIR=/Users/alex/tmp/gems/2.3.5/extensions/x86_64-darwin-16/2.3.0/scrypt-3.0.5 RUBYLIBDIR=/Users/alex/tmp/gems/2.3.5/extensions/x86_64-darwin-16/2.3.0/scrypt-3.0.5
mkdir -p x86_64-darwin
/nix/store/hm9pigz389gfvr07bid05axzvj93pz1m-ruby-2.3.5/bin/ruby -I/Users/alex/.gem/ruby/2.3.0/gems/ffi-compiler-1.0.1/lib/ffi-compiler/fake_ffi -I../../lib/scrypt /Users/alex/.gem/ruby/2.3.0/gems/ffi-compiler-1.0.1/lib/ffi-compiler/exporter.rb ../../lib/scrypt/scrypt_ext.rb ./scrypt_ext.h
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/crypto_scrypt.o -c ./crypto_scrypt.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/crypto_scrypt_smix.o -c ./crypto_scrypt_smix.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/crypto_scrypt_smix_sse2.o -c ./crypto_scrypt_smix_sse2.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/insecure_memzero.o -c ./insecure_memzero.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/memlimit.o -c ./memlimit.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/scrypt_calibrate.o -c ./scrypt_calibrate.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/scrypt_ext.o -c ./scrypt_ext.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/scryptenc_cpuperf.o -c ./scryptenc_cpuperf.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/sha256.o -c ./sha256.c
clang -fexceptions -O -fno-omit-frame-pointer -fno-strict-aliasing -Wall -std=c99 -msse -msse2 -arch x86_64 -arch i386   -o x86_64-darwin/warnp.o -c ./warnp.c
clang -bundle -o x86_64-darwin/libscrypt_ext.bundle x86_64-darwin/crypto_scrypt.o x86_64-darwin/crypto_scrypt_smix.o x86_64-darwin/crypto_scrypt_smix_sse2.o x86_64-darwin/insecure_memzero.o x86_64-darwin/memlimit.o x86_64-darwin/scrypt_calibrate.o x86_64-darwin/scrypt_ext.o x86_64-darwin/scryptenc_cpuperf.o x86_64-darwin/sha256.o x86_64-darwin/warnp.o -fexceptions -arch x86_64 -arch i386
ld: warning: ld: warning: ignoring file /nix/store/lcnjkcr311np24c1xpksj37mav4w477n-clang-4.0.1/bin/../lib/clang/4.0.1/lib/darwin/libclang_rt.osx.a, missing required architecture i386 in file /nix/store/lcnjkcr311np24c1xpksj37mav4w477n-clang-4.0.1/bin/../lib/clang/4.0.1/lib/darwin/libclang_rt.osx.a (2 slices)ignoring file /nix/store/ppphs7gv2li296n1wdy0jq2kk6gwxhya-Libsystem-osx-10.11.6/lib/libSystem.dylib, file was built for x86_64 which is not the architecture being linked (i386): /nix/store/ppphs7gv2li296n1wdy0jq2kk6gwxhya-Libsystem-osx-10.11.6/lib/libSystem.dylib

Undefined symbols for architecture i386:
  "___assert_rtn", referenced from:
      _PBKDF2_SHA256 in sha256.o
  "___error", referenced from:
      _crypto_scrypt in crypto_scrypt.o
      __crypto_scrypt in crypto_scrypt.o
      _memtouse in memlimit.o
      _libcperciva_warn in warnp.o
  "___stack_chk_fail", referenced from:
      _crypto_scrypt in crypto_scrypt.o
      _salsa20_8 in crypto_scrypt_smix.o
      _memtouse in memlimit.o
      _calibrate in scrypt_calibrate.o
      _scryptenc_cpuperf in scryptenc_cpuperf.o
      _libcperciva_SHA256_Update in sha256.o
      _libcperciva_SHA256_Final in sha256.o
      ...
  "___stack_chk_guard", referenced from:
      _crypto_scrypt in crypto_scrypt.o
      _salsa20_8 in crypto_scrypt_smix.o
      _memtouse in memlimit.o
      _calibrate in scrypt_calibrate.o
      _scryptenc_cpuperf in scryptenc_cpuperf.o
      _libcperciva_SHA256_Update in sha256.o
      _libcperciva_SHA256_Final in sha256.o
      ...
      _crypto_scrypt in crypto_scrypt.o
      _salsa20_8 in crypto_scrypt_smix.o
      _memtouse in memlimit.o
      _calibrate in scrypt_calibrate.o
      _scryptenc_cpuperf in scryptenc_cpuperf.o
      _libcperciva_SHA256_Update in sha256.o
      _libcperciva_SHA256_Final in sha256.o
      ...
      _crypto_scrypt in crypto_scrypt.o
      _salsa20_8 in crypto_scrypt_smix.o
      _memtouse in memlimit.o
      _calibrate in scrypt_calibrate.o
      _scryptenc_cpuperf in scryptenc_cpuperf.o
      _libcperciva_SHA256_Update in sha256.o
      _libcperciva_SHA256_Final in sha256.o
      ...
      _crypto_scrypt in crypto_scrypt.o
      _salsa20_8 in crypto_scrypt_smix.o
      _memtouse in memlimit.o
      _calibrate in scrypt_calibrate.o
      _scryptenc_cpuperf in scryptenc_cpuperf.o
      _libcperciva_SHA256_Update in sha256.o
      _libcperciva_SHA256_Final in sha256.o
      ...
      _crypto_scrypt in crypto_scrypt.o
      _salsa20_8 in crypto_scrypt_smix.o
      _memtouse in memlimit.o
      _calibrate in scrypt_calibrate.o
      _scryptenc_cpuperf in scryptenc_cpuperf.o
      _libcperciva_SHA256_Update in sha256.o
      _libcperciva_SHA256_Final in sha256.o
      ...
      _crypto_scrypt in crypto_scrypt.o
      _salsa20_8 in crypto_scrypt_smix.o
      _memtouse in memlimit.o
      _calibrate in scrypt_calibrate.o
      _scryptenc_cpuperf in scryptenc_cpuperf.o
      _libcperciva_SHA256_Update in sha256.o
      _libcperciva_SHA256_Final in sha256.o
      ...
      _crypto_scrypt in crypto_scrypt.o
      _salsa20_8 in crypto_scrypt_smix.o
      _memtouse in memlimit.o
      _calibrate in scrypt_calibrate.o
      _scryptenc_cpuperf in scryptenc_cpuperf.o
      _libcperciva_SHA256_Update in sha256.o
      _libcperciva_SHA256_Final in sha256.o
      ...
      ...
  "___stderrp", referenced from:
      _libcperciva_warn in warnp.o
      _libcperciva_warnx in warnp.o
  "_abort", referenced from:
      _crypto_scrypt in crypto_scrypt.o
  "_atexit", referenced from:
      _warnp_setprogname in warnp.o
  "_fprintf", referenced from:
      _libcperciva_warn in warnp.o
  "_fputs$UNIX2003", referenced from:
      _libcperciva_warn in warnp.o
      _libcperciva_warnx in warnp.o
  "_free", referenced from:
      __crypto_scrypt in crypto_scrypt.o
      _warnp_setprogname in warnp.o
      _done in warnp.o
  "_fwrite$UNIX2003", referenced from:
      _libcperciva_warn in warnp.o
      _libcperciva_warnx in warnp.o
  "_getrlimit$UNIX2003", referenced from:
      _memtouse in memlimit.o
  "_gettimeofday", referenced from:
      _scryptenc_cpuperf in scryptenc_cpuperf.o
  "_malloc", referenced from:
      __crypto_scrypt in crypto_scrypt.o
  "_memcmp", referenced from:
      _crypto_scrypt in crypto_scrypt.o
  "_memcpy", referenced from:
      __SHA256_Update in sha256.o
      __SHA256_Final in sha256.o
      _PBKDF2_SHA256 in sha256.o
  "_mmap$UNIX2003", referenced from:
      __crypto_scrypt in crypto_scrypt.o
  "_munmap$UNIX2003", referenced from:
      __crypto_scrypt in crypto_scrypt.o
  "_strdup", referenced from:
      _warnp_setprogname in warnp.o
  "_strerror$UNIX2003", referenced from:
      _libcperciva_warn in warnp.o
  "_strlen", referenced from:
      _crypto_scrypt in crypto_scrypt.o
  "_sysconf", referenced from:
      _memtouse in memlimit.o
  "_vfprintf", referenced from:
      _libcperciva_warn in warnp.o
      _libcperciva_warnx in warnp.o
ld: symbol(s) not found for architecture i386
clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
rake aborted!
Command failed with status (1): [clang -bundle -o x86_64-darwin/libscrypt_e...]
/Users/alex/.gem/ruby/2.3.0/gems/ffi-compiler-1.0.1/lib/ffi-compiler/compile_task.rb:153:in `block in define_task!'
/Users/alex/tmp/gems/2.3.5/gems/rake-12.3.0/exe/rake:27:in `<main>'
Tasks: TOP => default => x86_64-darwin/libscrypt_ext.bundle
(See full trace by running task with --trace)

rake failed, exit code 1

Gem files will remain installed in /Users/alex/tmp/gems/2.3.5/gems/scrypt-3.0.5 for inspection.
Results logged to /Users/alex/tmp/gems/2.3.5/extensions/x86_64-darwin-16/2.3.0/scrypt-3.0.5/gem_make.out
tracer:tmp alex$
```

It's interesting that the following works though:

```shell
tracer:foo alex$ GEM_HOME=$(pwd)/gems nix-shell --packages ruby_2_3

[nix-shell:~/tmp/foo]$ gem install scrypt
Fetching: rake-12.3.0.gem (100%)
Successfully installed rake-12.3.0
Fetching: ffi-1.9.18.gem (100%)
Building native extensions.  This could take a while...
Successfully installed ffi-1.9.18
Fetching: ffi-compiler-1.0.1.gem (100%)
Successfully installed ffi-compiler-1.0.1
Fetching: scrypt-3.0.5.gem (100%)
Building native extensions.  This could take a while...
Successfully installed scrypt-3.0.5
Parsing documentation for rake-12.3.0
Installing ri documentation for rake-12.3.0
Parsing documentation for ffi-1.9.18
Installing ri documentation for ffi-1.9.18
Parsing documentation for ffi-compiler-1.0.1
Installing ri documentation for ffi-compiler-1.0.1
Parsing documentation for scrypt-3.0.5
Installing ri documentation for scrypt-3.0.5
Done installing documentation for rake, ffi, ffi-compiler, scrypt after 18 seconds
4 gems installed
```

veprbl commented:
@aflatter
Thank you. I could *not* reproduce exactly your error on my system, but I think their build breaks because they want to compile with  ```-arch i386``` and this PR could fix that. On the other hand I find it suspicious that it works for you when you use interactive shell (notably you then drop the ```export PATH="$GEM_HOME/bin:$PATH"```). Perhaps there is an impurity in your process that causes the overall problem. I think the way to solve your problem would be to try to implement ```scrypt``` as a nix package, perhaps with some hacks to disable i386 build (I assume that is not crucial for it to work properly).

veprbl commented on pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix:
```diff
@@ -89,14 +99,24 @@ appleDerivation rec {
+       -reexport_library $out/lib/system/libsystem_kernel.i386.dylib \
+       $args
+
+    lipo -create $out/lib/libSystem.B.x86_64.dylib $out/lib/libSystem.B.i386.dylib -output $out/lib/libSystem.B.dylib
```
This doesn't play very well with ```stdenv/darwin/make-bootstrap-tools.nix```. In its ```build``` phase it will not pack ```libSystem.B.dylib```, instead remakes during the ```unpack``` phase. The problem is that other bootstrap tools binaries will depend on ```libSystem.B.x86_64.dylib``` which doesn't exist on the real darwin system. I guess I need to figure out a way to get rid of ```$out/lib/libSystem.B.x86_64.dylib``` and ```$out/lib/libSystem.B.i386.dylib``` which are needed when the gluing is done with ```lipo```.

copumpkin commented:
Any progress on this? I'm getting some trouble with smlnj and the lack of this

veprbl commented:
@copumpkin 
Unfortunately I am unable to bootstrap this so that we can compile an i386 version of libresolve and such. I was looking into splitting symbol list generation into a separate branch https://github.com/veprbl/nixpkgs/tree/libsystem_symbols, but I never managed to test it properly before I could submit a PR. The main problem is that the rebuilds are massive, I can't really do all of them on my laptop.
